PASO 1: Cargar el dataset en Google Colab, importa librerias

In [1]:
import pandas as pd
import numpy as np

# Cargar dataset
df = pd.read_csv('/content/sample_data/WA_Fn-UseC_-Telco-Customer-Churn.csv')

Validar la carga del data set

In [2]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [5]:
#Cantidad de Filas y coloumnas
df.shape

(7043, 21)

In [8]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [9]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

Validar Calidad de los datos

In [12]:
df.isnull().sum()



customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Validando o detectando anomalias

In [13]:
df['TotalCharges'].unique()[:20]


array(['29.85', '1889.5', '108.15', '1840.75', '151.65', '820.5',
       '1949.4', '301.9', '3046.05', '3487.95', '587.45', '326.8',
       '5681.1', '5036.3', '2686.05', '7895.15', '1022.95', '7382.25',
       '528.35', '1862.9'], dtype=object)

posibles clientes raros

In [14]:
df[df['TotalCharges'].str.strip() == ''][['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges']]


customerID  tenure  MonthlyCharges TotalCharges
488   4472-LVYGI       0           52.55             
753   3115-CZMZD       0           20.25             
936   5709-LVOEQ       0           80.85             
1082  4367-NUYAO       0           25.75             
1340  1371-DWPAZ       0           56.05             
3331  7644-OMVMY       0           19.85             
3826  3213-VVOLG       0           25.35             
4380  2520-SGTTA       0           20.00             
5218  2923-ARZLG       0           19.70             
6670  4075-WKNIU       0           73.35             
6754  2775-SEFEE       0           61.90

In [15]:
#Convertir TotalCharges a numérico
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')


In [16]:
df.isnull().sum()


customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [17]:
# estos son clientes nuevos que no tienen permanencia o
df[df['TotalCharges'].isnull()][['tenure', 'MonthlyCharges', 'Churn']].head(10)


tenure  MonthlyCharges Churn
488        0           52.55    No
753        0           20.25    No
936        0           80.85    No
1082       0           25.75    No
1340       0           56.05    No
3331       0           19.85    No
3826       0           25.35    No
4380       0           20.00    No
5218       0           19.70    No
6670       0           73.35    No

Proceder a colocar con ceros los socios que que tienen total charger null

In [18]:
df['TotalCharges'] = df['TotalCharges'].fillna(0)


In [19]:
#Validando
df.isnull().sum()


customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Distribución absoluta de la variable churn

In [20]:
df['Churn'].value_counts()


Churn
No     5174
Yes    1869
Name: count, dtype: int64

Distribucion porcentual

In [21]:
df['Churn'].value_counts(normalize=True) * 100


Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

Distribución de Churn

No: 5,174 clientes → 73.46%

Yes: 1,869 clientes → 26.54%

Explorar la Demografía vs Churn

In [26]:
pd.crosstab(df['gender'], df['Churn'], normalize='index') * 100


Churn          No        Yes
gender                      
Female  73.079128  26.920872
Male    73.839662  26.160338

🧠 Interpretación

Diferencia mínima (<1%)

👉 Género NO es un driver fuerte de churn

Esto es bueno: evita sesgos innecesarios

📌 Conclusión: variable débil, poco valor predictivo por sí sola.

SeniorCitizen vs Churn

In [23]:
pd.crosstab(df['SeniorCitizen'], df['Churn'], normalize='index') * 100


Churn                 No        Yes
SeniorCitizen                      
0              76.393832  23.606168
1              58.318739  41.681261

🧠 Interpretación

Los adultos mayores churnean casi el doble

Diferencia: +18 puntos

📌 Insight de negocio

Los clientes senior necesitan mejor soporte, planes más simples o atención diferenciada.

Partner vs Churn


In [24]:
pd.crosstab(df['Partner'], df['Churn'], normalize='index') * 100


Churn           No        Yes
Partner                      
No       67.042021  32.957979
Yes      80.335097  19.664903

🧠 Interpretación

Tener pareja reduce churn drásticamente

Estabilidad familiar = estabilidad como cliente

📌 Insight:

Clientes con pareja están más “anclados” al servicio.

Dependents vs Churn

In [25]:
pd.crosstab(df['Dependents'], df['Churn'], normalize='index') * 100


Churn              No        Yes
Dependents                      
No          68.720860  31.279140
Yes         84.549763  15.450237

🧠 Interpretación

Dependientes protegen aún más que la pareja

Clientes con hijos/personas a cargo valoran continuidad

📌 Insight fuerte:

Hogares familiares son los más leales.

🧠 Conclusión del PASO 5
Variable	Impacto
Gender	❌ Bajo
SeniorCitizen	🔥 Alto
Partner	🔥 Alto
Dependents	🔥 Muy alto

# Servicios vs Churn

## InternetService vs Churn

In [27]:
pd.crosstab(df['InternetService'], df['Churn'], normalize='index') * 100


Churn                   No        Yes
InternetService                      
DSL              81.040892  18.959108
Fiber optic      58.107235  41.892765
No               92.595020   7.404980

## Servicios de valor agregado

In [28]:
services = [
    'OnlineSecurity',
    'OnlineBackup',
    'DeviceProtection',
    'TechSupport',
    'StreamingTV',
    'StreamingMovies'
]

for col in services:
    print(f'\n{col}')
    print(pd.crosstab(df[col], df['Churn'], normalize='index') * 100)



OnlineSecurity
Churn                       No        Yes
OnlineSecurity                           
No                   58.233276  41.766724
No internet service  92.595020   7.404980
Yes                  85.388806  14.611194

OnlineBackup
Churn                       No        Yes
OnlineBackup                             
No                   60.071244  39.928756
No internet service  92.595020   7.404980
Yes                  78.468506  21.531494

DeviceProtection
Churn                       No        Yes
DeviceProtection                         
No                   60.872375  39.127625
No internet service  92.595020   7.404980
Yes                  77.497936  22.502064

TechSupport
Churn                       No        Yes
TechSupport                              
No                   58.364526  41.635474
No internet service  92.595020   7.404980
Yes                  84.833659  15.166341

StreamingTV
Churn                       No        Yes
StreamingTV                              
No

## **Servicios vs Churn (Interpretación)**
### 🔐 OnlineSecurity

Estado	Churn %

No	41.77% 🚨

Yes	14.61% ✅

No internet service	7.40%

🧠 Lectura

No tener seguridad casi triplica el churn

**Es uno de los drivers más fuertes**

Servicio altamente protector

### 💾 OnlineBackup

Estado	Churn %

No	39.93%

Yes	21.53%

No internet service	7.40%


📌 Menos fuerte que seguridad, pero sigue siendo clave.

### 📱 DeviceProtection

Estado	Churn %

No	39.13%

Yes	22.50%


👉 Protege, pero no tanto como seguridad o soporte.

### 🛠️ TechSupport (MUY IMPORTANTE)

Estado	Churn %

No	41.64% 🚨

Yes	15.17% ✅


**🔥 Segundo driver más fuerte junto a OnlineSecurity**

📌 Insight claro:

Clientes sin soporte técnico se van cuando algo falla.

### 📺 StreamingTV / Movies

Servicio	Churn % (Yes)

StreamingTV	~30%

StreamingMovies	~30%


🧠 Interpretación:

Poco impacto real

Son servicios “nice to have”

No retienen clientes

###🌐 “No internet service”

| Churn % | 7.40% |

👉 Clientes sin internet:

Servicio **simple**

Factura baja

Muy estables

## **Contrato y pagos (EL DRIVER #1)**




Contract vs Churn

In [29]:
pd.crosstab(df['Contract'], df['Churn'], normalize='index') * 100


Churn                  No        Yes
Contract                            
Month-to-month  57.290323  42.709677
One year        88.730482  11.269518
Two year        97.168142   2.831858

PaymentMethod vs Churn

In [30]:
pd.crosstab(df['PaymentMethod'], df['Churn'], normalize='index') * 100


Churn                             No        Yes
PaymentMethod                                  
Bank transfer (automatic)  83.290155  16.709845
Credit card (automatic)    84.756899  15.243101
Electronic check           54.714588  45.285412
Mailed check               80.893300  19.106700

PaperlessBilling vs Churn

In [31]:
pd.crosstab(df['PaperlessBilling'], df['Churn'], normalize='index') * 100


Churn                    No        Yes
PaperlessBilling                      
No                83.669916  16.330084
Yes               66.434908  33.565092

## Contrato y Pagos (worst offender #1)
📑 Contract vs Churn (worst offender)
Contrato	Churn %

**Month-to-month	42.71% 🚨🚨🚨**

One year	11.27%

Two year	2.83% ✅

🧠 Interpretación

Contrato mensual = cliente sin compromiso

Contrato 2 años = cliente casi imposible de perder

📌 Insight clave:

El churn NO es aleatorio, está estructuralmente ligado al tipo de contrato.

## 💳 PaymentMethod vs Churn

| Método               | Churn %       |
| -------------------- | ------------- |
| Electronic check     | **45.29%** 🚨 |
| Mailed check         | 19.11%        |
| Bank transfer (auto) | 16.71%        |
| Credit card (auto)   | 15.24%        |

🧠 Interpretación

Electronic check es altísimo riesgo

Pagos automáticos = retención

📌 Insight:

Fricción en el pago = churn.


## PaperlessBilling vs Churn

| Facturación electrónica | Churn %    |
| ----------------------- | ---------- |
| Yes                     | **33.57%** |
| No                      | 16.33%     |

🧠 Interpretación:

Paperless billing no es causa directa

Está correlacionada con:

Contrato mensual

Electronic check


## impacto por variables

| Variable         | Impacto        |
| ---------------- | -------------- |
| Contract         | 🔥🔥🔥🔥       |
| PaymentMethod    | 🔥🔥🔥         |
| PaperlessBilling | 🔥 (indirecto) |




# Variables numéricas

Tenure vs Churn

In [32]:
df.groupby('Churn')['tenure'].describe()


count       mean        std  min   25%   50%   75%   max
Churn                                                           
No     5174.0  37.569965  24.113777  0.0  15.0  38.0  61.0  72.0
Yes    1869.0  17.979133  19.531123  1.0   2.0  10.0  29.0  72.0

MonthlyCharges vs Churn

In [33]:
df.groupby('Churn')['MonthlyCharges'].describe()


count       mean        std    min    25%     50%   75%     max
Churn                                                                  
No     5174.0  61.265124  31.092648  18.25  25.10  64.425  88.4  118.75
Yes    1869.0  74.441332  24.666053  18.85  56.15  79.650  94.2  118.35

TotalCharges vs Churn

In [34]:
df.groupby('Churn')['TotalCharges'].describe()


count         mean          std    min    25%       50%      75%  \
Churn                                                                      
No     5174.0  2549.911442  2329.954215   0.00  572.9  1679.525  4262.85   
Yes    1869.0  1531.796094  1890.822994  18.85  134.5   703.550  2331.30   

           max  
Churn           
No     8672.45  
Yes    8684.80

# Variables numéricas (Interpretación)



## ⏳ Tenure vs Churn (EL FACTOR TEMPORAL)

| Métrica | No Churn       | Churn          |
| ------- | -------------- | -------------- |
| Media   | **37.6 meses** | **18.0 meses** |
| Mediana | **38 meses**   | **10 meses**   |
| Q1      | 15             | 2              |


🧠 Lectura

Clientes que se van:

Son mucho más nuevos

La mitad se va antes del mes 10

Retención ocurre después de los primeros meses críticos

📌 Insight:

Si sobrevives el primer año, es muy probable que te quedes.

# 💰 MonthlyCharges vs Churn
| Métrica | No    | Yes       |
| ------- | ----- | --------- |
| Media   | 61.27 | **74.44** |
| Mediana | 64.43 | **79.65** |



🧠 Lectura

Churn = facturas más altas

Clientes sienten que no vale lo que pagan

📌 Insight:

Precio alto sin servicios críticos = abandono.

# 🧾 TotalCharges vs Churn

| Métrica | No          | Yes        |
| ------- | ----------- | ---------- |
| Mediana | **1679.53** | **703.55** |


🧠 Lectura

TotalCharges refleja tenure

Clientes churn:

Poco tiempo

Poco valor acumulado

👉 Variable útil, pero redundante con tenure.

🧠 Resumen
🔥 Principales drivers de churn

1️⃣ Contract (Month-to-month)

2️⃣ Electronic check

3️⃣ No TechSupport / No OnlineSecurity

4️⃣ SeniorCitizen

5️⃣ Tenure bajo (< 12 meses)

6️⃣ MonthlyCharges altos

🎯 Segmento de mayor riesgo

Clientes nuevos, con contrato mensual,
factura alta, sin soporte ni seguridad,
que pagan con electronic check.

💡 Acciones de negocio sugeridas

Migrar clientes mensuales a contratos anuales

Bundles con TechSupport + OnlineSecurity

Incentivos en los primeros 3–6 meses

Programas específicos para seniors

Eliminar fricción en pagos